### PR TITLE
MRG: Annotate return type of `data_path()` functions

### PR DESCRIPTION
**Pull request by hoechenberger**
I often ran into the problem that Pylance couldn't correctly infer the return type of the `data_path()` functions for our datasets. Let's look at the following MWE.

```python
# %%
from typing import reveal_type

import mne

reveal_type(mne.datasets.sample.data_path(download=False))
```

### This is what we get on `main`:

```shell
❯ pyright mwe.py
/private/tmp/mwe.py
  /private/tmp/mwe.py:6:13 - information: Type of "mne.datasets.sample.data_path(download=False)" is "Unknown"
0 errors, 0 warnings, 1 information 
❯ python mwe.py
Runtime type is 'PosixPath'
```

### Now, on this PR branch we get:
```shell
❯ pyright mwe.py
/private/tmp/mwe.py
  /private/tmp/mwe.py:6:13 - information: Type of "mne.datasets.sample.data_path(download=False)" is "Path"
0 errors, 0 warnings, 1 information 
❯ python mwe.py
Runtime type is 'PosixPath'
```
To achieve this improved behavior, I had to use overloads – function signatures that basically cover the different values of the `return_version` parameter and the associated return type.

Some might not like this because we're essentially duplicating (actually, triplicating) the function signature. My argument **for** doing this change are:

- Improved user experience
- No runtime implications – if the signatures get out of sync, no user code will break
- Trivial maintenance, since no function implementation is multiplied

I'd volunteer to be responsible for this part of the code if it breaks.

We can also move the overloads to a `.pyi` file so they're not visually clogging the Python module, in case that's a concern.

I'd like to hear your thoughts!

### CODE REVIEW

File: doc/changes/devel/12734.newfeature.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+When using the ``data_path()`` in any dataset included in with :py:mod:`mne.datasets`,
```
Comment (hoechenberger) [inline doc/changes/devel/12734.newfeature.rst:1]:
```suggestion
When using the ``data_path()`` in any dataset included in :py:mod:`mne.datasets`,
```

**Comment by cbrnr** [comment]
I'm OK with this change. Minor question, `PosixPath` inherits from `Path`, right? Is it a problem (for you) that the statically inferred type still differs from the runtime type?

**Comment by hoechenberger** [comment]
> I'm OK with this change. Minor question, `PosixPath` inherits from `Path`, right? Is it a problem (for you) that the statically inferred type still differs from the runtime type?

I was asking myself the same question, and the class hierarchy I found in the Python docs actually says it's the other way around – `Path` inherits from `PosixPath`:




Which I found confusing; but then I found this:

> class pathlib.Path(*pathsegments)
> A subclass of [PurePath](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath), this class represents concrete paths of the system’s path flavour (instantiating it creates either a [PosixPath](https://docs.python.org/3/library/pathlib.html#pathlib.PosixPath) or a [WindowsPath](https://docs.python.org/3/library/pathlib.html#pathlib.WindowsPath)):

So I think we're good!

**Comment by drammock** [comment]
I personally don't like this. But

> * Improved user experience
> * No runtime implications – if the signatures get out of sync, no user code will break
> * Trivial maintenance, since no function implementation is multiplied

I guess that's hard to argue with. If the signatures do get out of sync though, then "improved user experience" stops happening right?  In other words, there is a (small) maintenance burden to retaining the usability advantage of this change?

Putting the overloads in a `.pyi` file seems like *inviting* them to get out of sync, so I'm not sure that's a good choice.

Is there no more concise way to achieve the same end (now or in the near future)?  E.g. is there a package that makes this easier, or maybe "this gets much cleaner once Python 3.14 is out" or something?  (sorry for the naive questions, as you know I'm not really on top of the python type annotations state of the art)

**Comment by cbrnr** [comment]
> Is there no more concise way to achieve the same end

Usually, this would be solved with a `Union`, but I guess here it is not possible and/or would be even more cumbersome?

File: mne/datasets/_fetch.py [hunk 2]
```diff
@@ -42,7 +75,7 @@ def fetch_dataset(
     accept=False,
     auth=None,
     token=None,
-):
+) -> Union[Path, tuple[Path, str]]:
```
Comment (drammock) [inline mne/datasets/_fetch.py:78]: @cbrnr `Union` is used here, but I'm inferring that it isn't enough by itself

Comment (cbrnr) [reply mne/datasets/_fetch.py:78]: Yes, the problem is that the return type depends on the argument, so I guess there's no way around `@overload`

**Comment by hoechenberger** [comment]
@cbrnr 
> Usually, this would be solved with a `Union`, but I guess here it is not possible and/or would be even more cumbersome?

We **are** using a `Union` (one has to in this case). Without  the overloads, this only slightly helps our users though: They'd have to use `assert` statements as type guards to make the type checker aware of which return values to expect.

> I guess that's hard to argue with. If the signatures do get out of sync though, then "improved user experience" stops happening right? In other words, there is a (small) maintenance burden to retaining the usability advantage of this change?

Yes, but it's something that would be could very quickly (suddenly: squiggly red lines where there used not to be any).

> Putting the overloads in a `.pyi` file seems like _inviting_ them to get out of sync, so I'm not sure that's a good choice.

I agree, I'm not a fan of `.pyi` files for this reason.

> Is there no more concise way to achieve the same end

Unfortunately I couldn't find any for this use case, no. I explored `TypeGuard` and `TypeVar` and couldn't come up with any working solution (tested with upright and Pylance)

File: mne/datasets/_fetch.py (hunk 2 above)
Comment (hoechenberger) [reply mne/datasets/_fetch.py:78]: Annotating only the `Union` and omitting the overloads makes it impossible for the type checker to infer which of the two possible return types (in this specific case, Path or tuple of Path and string) to expect based on the parameters provided to the function

**Comment by hoechenberger** [comment]
I have another idea which requires more lines to change but avoids the overloads.

Instead of annotating this rather generic `fetch_dataset()` function, we could annotate the `data_path()` functions directly -- they **always** return a Path -- and are probably the more important parts of the API anyway (from a user's perspective)

WDYT?

**Comment by cbrnr** [comment]
I was going to say that a function that returns different types depending on one or more arguments is perhaps not ideal, but this would be a more invasive change. If you think annotating data_path functions is sufficient, I'd say go for it! I have to agree with @drammock that having all that stuff in .pyi files is not great, but unfortunately our adoption of lazy load forces us to do that. Or does it? Can we not annotate directly in .py sources for the type checker to pick it up?

**Comment by hoechenberger** [comment]
I'll give it a shot 

regarding pyi files, yes they're optional

**Comment by hoechenberger** [comment]
@drammock @cbrnr I figured I could also just annotate the `_download_mne_dataset()` function, which **also** always returns a Path! Much simpler and it yields the same user-visible effect

**Review by drammock — APPROVED** [review]
Nice, much cleaner!

File: mne/datasets/utils.py [hunk 3]
```diff
@@ -252,6 +252,8 @@ def _download_mne_dataset(
         download=download,
         accept=accept,
     )
+    assert isinstance(path, Path)
+    return path
```
Comment (cbrnr) [inline mne/datasets/utils.py:256]: Is this really necessary? For me, the return type annotation is sufficient for the type checker to infer the correct return type.

Comment (hoechenberger) [reply mne/datasets/utils.py:256]: Yes but without this type guard, the return value here in this fu drin will get red squiggly lines. This is of course not user-facing but something that drives me mad as a developer, so I added the assert statement to take care of it

File: mne/datasets/utils.py [hunk 4]
```diff
@@ -252,6 +252,8 @@ def _download_mne_dataset(
         download=download,
         accept=accept,
     )
+    assert isinstance(path, Path)
```
Comment (hoechenberger) [inline mne/datasets/utils.py:255]:
```suggestion
    assert isinstance(path, Path)  # help the type checker
```

**Comment by hoechenberger** [comment]
@cbrnr I've added a comment to explain the assert statement

**Comment by hoechenberger** [comment]
@cbrnr if the assert is a problem then we'll have to revert to using overloads

I'm ok with either approach

**Comment by hoechenberger** [comment]
I didn't know about the optimization flag implications 

then again, we don't need to do anything just the type checker at runtime, since the analysis is carried-out offline

**Comment by cbrnr** [comment]
> then again, we don't need to do anything just the type checker at runtime, since the analysis is carried-out offline

Sorry, I don't understand what you're saying here...

**Comment by hoechenberger** [comment]
i bitched the sentence 😅

i meant to say:

even if the assert is removed during optimization -- it wouldn't matter in this particular case as the assert is really just here for the static code analysis tools and is not actually intended to have any runtime implications 

one could emphasize this even more strongly by putting it into an `if typing.TYPE_CHECKING` block

**Comment by cbrnr** [comment]
Got it 😄! My main issue is not the assert statement though (I think this would be fine together with a comment that you already added), but the fact that in order to assert the type we now need to name the function call return value and then return it, which is ugly. Like I said, in this case I'd prefer an overload, but I'm sure that similar situations will crop up in the future where such an ugly construct is the only viable solution 😢...

**Comment by hoechenberger** [comment]
@cbrnr I might be able to "force" the type checker to assume a certain type by using typing.cast

one could then do:

return cast(TargetClass, my_func())

cast would have no runtime implications 

it feels a bit dirty but maybe for this particular case it could be the best compromise

**Comment by cbrnr** [comment]
Yes, I don't think there is a "beautiful" solution here, just solutions with varying degrees of "ugliness". I can really live with all variants, because they improve user experience dramatically, so I'll leave the decision up to you.

**Comment by hoechenberger** [comment]
@cbrnr @drammock 
I'm now using [`typing.cast`](https://docs.python.org/3/library/typing.html#typing.cast) to avoid any runtime implications. **If** anything breaks, it will only affect type checkers / static code analysis tools; no user scripts will be harmed. :)

File: mne/datasets/_fetch.py [hunk 5]
```diff
@@ -3,6 +3,8 @@
 # License: BSD-3-Clause
 # Copyright the MNE-Python contributors.
 
+from __future__ import annotations  # Only needed for Python 3.9
```
Comment (cbrnr) [inline mne/datasets/_fetch.py:6]:
```suggestion
from __future__ import annotations  # only needed for Python ≤ 3.9
```

**Comment by larsoner** [comment]
In the future, I don't mind the `assert isinstance(...)` approach either if it's more convenient (one fewer import generally at least). But the `cast` seems okay, too, so I'll go ahead and merge, thanks @hoechenberger !

**Comment by drammock** [comment]
> I don't mind the `assert isinstance(...)` approach either if it's more convenient (one fewer import generally at least). But the `cast` seems okay, too

same here, FWIW

**Pull request merged**